docs: Update LayoutLM model card according to new standardized format

Jin-HoMLee opened the pull request:
# Update LayoutLM Model Card Documentation

This PR updates the LayoutLM model card documentation according to the standardized format as issued in https://github.com/huggingface/transformers/issues/36979. The changes improve the documentation's clarity and usability while maintaining consistency with other model cards in the repository. 

Changes include: 

- A badge indicating model support (PyTorch, TensorFlow) in the upper right side of the doc
- A brief description of the model (what makes it unique/different) written in a way that's accessible to everyone
- Ready to use example code 
    - featuring AutoModel
    - featuring quantization using TorchAoConfig
- Short explanations that Pipeline and transformers-cli are not supported
- Notes 
    - that additional pre-processing steps are needed with link to further resources
    - and reference on the improved, most recent LayoutLMv3 model
    - that attention mask visualizer is not supported

The changes make the documentation more accessible and provide ready-to-use examples for different use cases, following the standardized format used in other model cards like Gemma 3, PaliGemma, and ViT.


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Since this is a documentation update for a vision-language model, I would suggest tagging:

@amyeroberts (vision models)
@stevhliu (documentation)

Jin-HoMLee committed:
docs: Update LayoutLM model card with standardized format

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -14,51 +14,120 @@ rendered properly in your Markdown viewer.
+You can find all the original LayoutLM checkpoints under the [LayoutLM](https://huggingface.co/collections/microsoft/layoutlm-6564539601de72cb631d0902) collection.
+
+> [!TIP]
+> This model was contributed by [microsoft](https://github.com/microsoft).
+>
```
```suggestion
```

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -14,51 +14,120 @@ rendered properly in your Markdown viewer.
+>
+> Click on the LayoutLM models in the right sidebar for more examples of how to apply LayoutLM to different vision and language tasks.
+
+The example below demonstrates how to use the model with the ([`Pipeline`](https://huggingface.co/docs/transformers/main/en/main_classes/pipelines#transformers.Pipeline) or) [`AutoModel`](https://huggingface.co/docs/transformers/main/en/model_doc/auto#transformers.AutoModel) class. 
```
```suggestion
The example below demonstrates question answering with the [`AutoModel`] class. 
```

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
````diff
@@ -14,51 +14,120 @@ rendered properly in your Markdown viewer.
+</hfoption>
+<hfoption id="AutoModel">
+
+```py
````
More helpful to show a document answering example

```py
import torch
from datasets import load_dataset
from transformers import AutoTokenizer, LayoutLMForQuestionAnswering

tokenizer = AutoTokenizer.from_pretrained("impira/layoutlm-document-qa", add_prefix_space=True)
model = LayoutLMForQuestionAnswering.from_pretrained("impira/layoutlm-document-qa", torch_dtype=torch.float16)

dataset = load_dataset("nielsr/funsd", split="train")
example = dataset[0]
question = "what's his name?"
words = example["words"]
boxes = example["bboxes"]

encoding = tokenizer(
    question.split(), words, is_split_into_words=True, return_token_type_ids=True, return_tensors="pt"
)
bbox = []
for i, s, w in zip(encoding.input_ids[0], encoding.sequence_ids(0), encoding.word_ids(0)):
    if s == 1:
        bbox.append(boxes[w])
    elif i == tokenizer.sep_token_id:
        bbox.append([1000] * 4)
    else:
        bbox.append([0] * 4)
encoding["bbox"] = torch.tensor([bbox])

word_ids = encoding.word_ids(0)
outputs = model(**encoding)
loss = outputs.loss
start_scores = outputs.start_logits
end_scores = outputs.end_logits
start, end = word_ids[start_scores.argmax(-1)], word_ids[end_scores.argmax(-1)]
print(" ".join(words[start : end + 1]))
```

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -14,51 +14,120 @@ rendered properly in your Markdown viewer.
-  position of the lower right corner. Note that one first needs to normalize the bounding boxes to be on a 0-1000
-  scale. To normalize, you can use the following function:
+</hfoption>
+<hfoption id="transformers-cli">
```
This can be removed as well

Jin-HoMLee replied:
I don't know why the part about bounding normalization got here..., but I guess you mean that I can delete the part about transformer-cli, right? :)

stevhliu replied:
Yeah I was referring to the transformers CLI part!

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -14,51 +14,120 @@ rendered properly in your Markdown viewer.
+</hfoption>
+</hfoptions>
+
+Quantization reduces the memory burden of large models by representing the weights in a lower precision. 
```
The model isn't that large so no need to include a quantization example

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -14,51 +14,120 @@ rendered properly in your Markdown viewer.
+
+## Notes
+
+- The original LayoutLM was not designed with a unified processing workflow. Instead, it expects you to pre-process text (`words`) and bounding boxes (`boxes`) via an external OCR engine, e.g. Google's Tesseract (there’s a [Python wrapper](https://pypi.org/project/pytesseract/) available) and provide them as additional inputs to the tokenizer. 
```
```suggestion
- The original LayoutLM was not designed with a unified processing workflow. Instead, it expects preprocessed text (`words`) and bounding boxes (`boxes`) from an external OCR engine (like [Pytesseract](https://pypi.org/project/pytesseract/)) and provide them as additional inputs to the tokenizer. 
```

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -14,51 +14,120 @@ rendered properly in your Markdown viewer.
+## Notes
+
+- The original LayoutLM was not designed with a unified processing workflow. Instead, it expects you to pre-process text (`words`) and bounding boxes (`boxes`) via an external OCR engine, e.g. Google's Tesseract (there’s a [Python wrapper](https://pypi.org/project/pytesseract/) available) and provide them as additional inputs to the tokenizer. 
+For simplified demonstration purposes, the above code examples use example data from the `datasets` library, which come along with already pre-processed words and boxes. 
+
+- For more details and examples of usage refer to the links in the [Resources](#resources) section. 
```
```suggestion
```

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -14,51 +14,120 @@ rendered properly in your Markdown viewer.
+
+- For more details and examples of usage refer to the links in the [Resources](#resources) section. 
+
+- **Bounding Boxes**: Bounding box coordinates should be in the format `(x_min, y_min, x_max, y_max)`. It's best practice to normalize these coordinates to a range like `[0, 1000]`. To normalize, you can use the following function:
```
```suggestion
- The [`~LayoutLM.forward`] method expects the input `bbox` (bounding boxes of the input tokens). Each bounding box should be in the format `(x0, y0, x1, y1)`.  `(x0, y0)` corresponds to the upper left corner of the bounding box and `{x1, y1)` corresponds to the lower right corner. The bounding boxes need to be normalized on a 0-1000 scale as shown below.
```

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -70,8 +139,7 @@ def normalize_bbox(bbox, width, height):
 
-Here, `width` and `height` correspond to the width and height of the original document in which the token
-occurs. Those can be obtained using the Python Image Library (PIL) library for example, as follows:
+- `width` and `height` correspond to the width and height of the original document in which the token occurs. Those can be obtained using the Python Image Library (PIL) library for example, as follows:
```
```suggestion
- `width` and `height` correspond to the width and height of the original document in which the token occurs. These values can be obtained as shown below.
```

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -82,99 +150,106 @@ image = Image.open(name_of_your_document).convert("RGB")
+- **Tokenization**: The model uses the standard `BertTokenizer` to tokenize the text (, and you'll need to pass the bounding box information for each token). 
 
-🚀 Deploy
+- [**AttentionMaskVisualizer**](https://github.com/huggingface/transformers/blob/beb9b5b02246b9b7ee81ddf938f93f44cfeaad19/src/transformers/utils/attention_visualizer.py#L139) is not directly compatible with LayoutLM's multi-modal output (text and layout) and attention structure (as opposed to text-based models like BERT with standard attention mechanisms).  
```
Remove the `AttentionMaskVisualizer` section since its unsupported

Jin-HoMLee replied:
Thanks, done.

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -82,99 +150,106 @@ image = Image.open(name_of_your_document).convert("RGB")
+- [**AttentionMaskVisualizer**](https://github.com/huggingface/transformers/blob/beb9b5b02246b9b7ee81ddf938f93f44cfeaad19/src/transformers/utils/attention_visualizer.py#L139) is not directly compatible with LayoutLM's multi-modal output (text and layout) and attention structure (as opposed to text-based models like BERT with standard attention mechanisms).  
 
-- A blog post on how to [Deploy LayoutLM with Hugging Face Inference Endpoints](https://www.philschmid.de/inference-endpoints-layoutlm).
+- The AutoProcessor in [**LayoutLMv3**](https://huggingface.co/docs/transformers/en/model_doc/layoutlmv3) greatly simplifies the workflow by integrating all mentioned steps into a single, user-friendly class. 
```
```suggestion
```

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -82,99 +150,106 @@ image = Image.open(name_of_your_document).convert("RGB")
+- The AutoProcessor in [**LayoutLMv3**](https://huggingface.co/docs/transformers/en/model_doc/layoutlmv3) greatly simplifies the workflow by integrating all mentioned steps into a single, user-friendly class. 
 
-## LayoutLMConfig
+### LayoutLMConfig
```
Don't need to change any of the API section

Jin-HoMLee replied:
Great, thanks! :)

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
```diff
@@ -82,99 +150,106 @@ image = Image.open(name_of_your_document).convert("RGB")
-
-**Other resources**
-- [Masked language modeling task guide](../tasks/masked_language_modeling)
+- **Tokenization**: The model uses the standard `BertTokenizer` to tokenize the text (, and you'll need to pass the bounding box information for each token). 
```
```suggestion
```

stevhliu commented on docs/source/en/model_doc/layoutlm.md:
````diff
@@ -82,99 +150,106 @@ image = Image.open(name_of_your_document).convert("RGB")
 width, height = image.size
 ```
 
-## Resources
````
For the Resources section, let's keep the external ones. You can just use a simple list for them:

```md
- Read [fine-tuning LayoutLM for document-understanding using Keras & Hugging Face Transformers](https://www.philschmid.de/fine-tuning-layoutlm-keras) to learn more.
- Read [fine-tune LayoutLM for document-understanding using only Hugging Face Transformers](https://www.philschmid.de/fine-tuning-layoutlm) for more information.
- Refer to this [notebook](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Add_image_embeddings_to_LayoutLM.ipynb) for a practical example of how to fine-tune LayoutLM.
- Refer to this [notebook](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Fine_tuning_LayoutLMForSequenceClassification_on_RVL_CDIP.ipynb) for an example of how to fine-tune LayoutLM for sequence classification.
- Refer to this [notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Fine_tuning_LayoutLMForTokenClassification_on_FUNSD.ipynb) for an example of how to fine-tune LayoutLM for token classification.
- Read [Deploy LayoutLM with Hugging Face Inference Endpoints](https://www.philschmid.de/inference-endpoints-layoutlm) to learn how to deploy LayoutLM.
```

Jin-HoMLee replied:
Ok, done :)

stevhliu reviewed (commented):
Thanks for your work!

Jin-HoMLee committed:
Apply suggestions from code review

This commit incorporates all suggestions provided in the recent review. Further changes will be committed separately to address remaining comments.

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

Jin-HoMLee reviewed (commented):
Thank you so much @stevhliu for the quick review! 
I committed all your suggested changes in e6e205d9235f2b7b8d774491f4ed5554c01fa4c0. Just worked on the remaining comments. Commit will follow > bd7155462a759d5ff75ec84bf2601c97c73c0639. Cheers!

Jin-HoMLee committed:
Address remaining review comments

stevhliu reviewed (commented):
Thanks again, just a few more things!

1. Remove the transformers CLI section.
2. Put the `## Resources` section after `## Notes`.
3. The API references (`LayoutLMTokenizer`) should use a second-level header `##`

Jin-HoMLee committed:
Address few more review comments:
1. remove transformer-cli section
2. put resources after notes
3. change API refs to 2nd level header

stevhliu committed:
Update layoutlm.md

stevhliu committed:
Update layoutlm.md

stevhliu reviewed (approved):
LGTM! 👍
